lib/livebook/runtime/evaluator.ex

```diff
@@ -689,25 +695,169 @@ defmodule Livebook.Runtime.Evaluator do
+    try do
+      with {:ok, form_statements} <- tokens_to_forms(tokens),
+           {:ok, forms} <- parse_forms(form_statements) do
+        # First statement - form = module definition
+        {:attribute, _, :module, module_name} = hd(forms)
+
+        # Compile the forms from the code-block
+        {:ok, _, binary_module} = :compile.forms(forms)
+
+        # Line below should be done in Tracer...
+        {:module, module_name} =
+          :code.load_binary(module_name, ~c"#{module_name}.beam", binary_module)
+
+        # Registration of module
+        # env = %{env | versioned_vars: %{}}
+        env = %{env | module: module_name}
+
+        Evaluator.Tracer.trace({:on_module, binary_module, %{}}, env)
+
+        {{:ok, ~c"erlang module successfully compiled", binding, env}, []}
+      else
+        # Tokenizer error - https://www.erlang.org/doc/man/erl_scan.html#string-3
+        {:error, {location, module, description}, _end_loc} ->
+          process_erlang_error(env, code, location, module, description)
+
+        # Parser error - https://www.erlang.org/doc/man/erl_parse.html#parse_form-1
+        {:error, {location, module, description}} ->
+          process_erlang_error(env, code, location, module, description)
+      end
+    catch
+      kind, error ->
+        stacktrace = prune_stacktrace(:erl_eval, __STACKTRACE__)
+        {{:error, kind, error, stacktrace}, []}
+    end
+  end
+
+  defp eval_statements(:erlang, code, tokens, binding, env) do
     try do
       erl_binding =
         Enum.reduce(binding, %{}, fn {name, value}, erl_binding ->
           :erl_eval.add_binding(elixir_to_erlang_var(name), value, erl_binding)
         end)
 
-      with {:ok, tokens, _} <- :erl_scan.string(String.to_charlist(code), {1, 1}, [:text]),
-           {:ok, parsed} <- :erl_parse.parse_exprs(tokens),
+      with {:ok, parsed} <- :erl_parse.parse_exprs(tokens),
            {:value, result, new_erl_binding} <- :erl_eval.exprs(parsed, erl_binding) do
         # Simple heuristic to detect the used variables. We look at
         # the tokens and assume all var tokens are used variables.
         # This will not handle shadowing of variables in fun definitions
         # and will only work well enough for expressions, not for modules.
+
         used_vars =
           for {:var, _anno, name} <- tokens,
               do: {erlang_to_elixir_var(name), nil},
               into: MapSet.new(),
               uniq: true
 
+        # Future idea - make sure to know which module we are calling from
+        # in order to have a 'Stale'-indication:
+        # called_functions =
+        #   for {:call, _, {:remote, _, {:atom, _, call_module}, {:atom, _, call_function}}, _} <-
+        #         parsed,
+        #       do: {{erlang_to_elixir_var(call_module), erlang_to_elixir_var(call_function)}, nil},
+        #       into: MapSet.new(),
+        #       uniq: true
+
         # Note that for Elixir we evaluate with :prune_binding, here
```

Comment:
You should also remove the "# Tokenize error" branch from the else below, as we are doing the scanning higher. :)